### PR TITLE
Prefer language settings #87

### PR DESCRIPTION
**Pull request by PangMo5**
## Changed
- Add prefer language settings
- Apply prefer language settings to iOS VideoPlayer
- Add https://github.com/sindresorhus/Defaults Package  
	- If use it, reduces human issues with auto-completion. And more sugar!
- Replaced `UserDefaults.default` with `Defaults`

### CODE REVIEW

**Comment by stephenb10** [comment]
Would these options apply across all media? Is there anything sever side to have preferences apply to an entire series?

**Comment by PangMo5** [comment]
> Would these options apply across all media? Is there anything sever side to have preferences apply to an entire series, e.g a show to have a preference of no subtitles but prefer subtitles for everything else

Currently, the settings are supposed to be applied to the entire video.

File: JellyfinPlayer/VideoPlayer.swift [hunk 1]
```diff
@@ -558,11 +558,7 @@ class PlayerViewController: UIViewController, GCKDiscoveryManagerListener, GCKRe
         }
         
         audioTrackArray.forEach { audio in
-            if Defaults[.autoSelectAudioLangCode] == "Auto",
-               audio.langCode.contains(Locale.current.languageCode ?? "") {
-                selectedAudioTrack = audio.id
-                mediaPlayer.currentAudioTrackIndex = audio.id
-            } else if audio.langCode.contains(Defaults[.autoSelectAudioLangCode]) {
+            if audio.langCode.contains(Defaults[.autoSelectAudioLangCode]) {
                 selectedAudioTrack = audio.id
```
Comment (stephenb10) [inline JellyfinPlayer/VideoPlayer.swift:562]: There is also a function to set either the audio or subtitle tracks

File: JellyfinPlayer/SettingsView.swift [hunk 2]
```diff
@@ -7,28 +7,30 @@
...
+    
     @ObservedObject var viewModel: SettingsViewModel
-
+    
     @Binding var close: Bool
-    @State private var inNetworkStreamBitrate: Int = 40_000_000
-    @State private var outOfNetworkStreamBitrate: Int = 40_000_000
-    @State private var autoSelectSubtitles: Bool = false
-    @State private var autoSelectSubtitlesLangcode: String = "none"
+    @Default(.inNetworkBandwidth)
```
Comment (acvigue) [inline JellyfinPlayer/SettingsView.swift:18]: make these one line

File: JellyfinPlayer/SettingsView.swift [hunk 3]
```diff
@@ -37,29 +39,35 @@ struct SettingsView: View {
...
-
+                
                 Section(header: Text("Accessibility")) {
-                    Toggle("Automatically show subtitles", isOn: $autoSelectSubtitles).onChange(of: autoSelectSubtitles, perform: { _ in
-                        let defaults = UserDefaults.standard
-                        defaults.setValue(autoSelectSubtitles, forKey: "AutoSelectSubtitles")
-                    })
-                    Picker("Language preferences", selection: $autoSelectSubtitlesLangcode) {}
+                    Toggle("Automatically show subtitles", isOn: $isAutoSelectSubtitles)
+                    SearchablePicker(label: "Subtitles language preferences",
```
Comment (acvigue) [inline JellyfinPlayer/SettingsView.swift:53]: "preferred subtitle language"

File: JellyfinPlayer/SettingsView.swift [hunk 4]
```diff
@@ -37,29 +39,35 @@ struct SettingsView: View {
...
+                                     options: viewModel.langs,
+                                     optionToString: { $0.name },
+                                     selected:Binding<Lang>(
+                                        get: { viewModel.langs.first(where: { $0.isoCode == autoSelectSubtitlesLangcode }) ?? .auto },
+                                        set: {autoSelectSubtitlesLangcode = $0.isoCode}
+                                     )
+                    )
+                    SearchablePicker(label: "Audio language preferences",
```
Comment (acvigue) [inline JellyfinPlayer/SettingsView.swift:61]: "preferred audio language"

File: JellyfinPlayer/VideoPlayer.swift [hunk 5]
```diff
@@ -543,7 +542,29 @@ class PlayerViewController: UIViewController, GCKDiscoveryManagerListener, GCKRe
                 .store(in: &cancellables)
         }
     }
-
+    
+    func setupTracksForPreferLanguage() {
```
Comment (acvigue) [inline JellyfinPlayer/VideoPlayer.swift:546]: not proper function naming.
"setupTrakcsForPreferredDefaults"

**Review by acvigue — CHANGES_REQUESTED** [review]

File: JellyfinPlayer/VideoPlayer.swift (hunk 5 above)
Comment (PangMo5) [reply JellyfinPlayer/VideoPlayer.swift:546]: tracks?

Comment (acvigue) [reply JellyfinPlayer/VideoPlayer.swift]: yes :p

File: Shared/ViewModels/SettingsViewModel.swift [hunk 6]
```diff
@@ -23,8 +23,17 @@ struct Bitrates: Codable, Hashable {
     public var value: Int
 }
 
+struct Lang: Hashable {
```
Comment (acvigue) [inline Shared/ViewModels/SettingsViewModel.swift:26]: probably change this to something such as `TrackLanguage`

File: Shared/ViewModels/VideoPlayerModel.swift [hunk 7]
```diff
@@ -16,10 +16,12 @@ struct Subtitle {
     var url: URL?
     var delivery: SubtitleDeliveryMethod
     var codec: String
+    var langCode: String
```
Comment (acvigue) [inline Shared/ViewModels/VideoPlayerModel.swift:19]: while short, langcode may cause confusion for new contributors. please change to `languageCode`

**Review by acvigue — CHANGES_REQUESTED** [review]

**Comment by PangMo5** [comment]
oh build error.. sry wait a min

**Review by acvigue — APPROVED** [review]

**Pull request merged**